File: clients/golang/v1/message.go
package v1

import (
	"encoding/json"
	"fmt"
)

// MsgStatus 对应 Rust 的 MsgStatus 枚举
type MsgStatus string

const (
	MsgStatusSuccess MsgStatus = "Success"
	MsgStatusWarning MsgStatus = "Warning"
	MsgStatusFailure MsgStatus = "Failure"
	MsgStatusError   MsgStatus = "Error"
)

// MessageType 定义所有消息类型
const (
	TypeReqPing              = "ReqPing"
	TypeRespPing             = "RespPing"
	TypeReqSubscribeTopic    = "ReqSubscribeTopic"
	TypeRespSubscribeTopic   = "RespSubscribeTopic"
	TypeUnsubscribeTopic     = "UnsubscribeTopic"
	TypeRespUnsubscribeTopic = "RespUnsubscribeTopic"
	TypeReqPublish           = "ReqPublish"
	TypeRespPublish          = "RespPublish"
	TypeReqSubscribe         = "ReqSubscribe"
	TypeRespSubscribe        = "RespSubscribe"
	TypeReqConsumerTopic     = "ReqConsumerTopic"
	TypeRespConsumerTopic    = "RespConsumerTopic"
	TypeReqUnConsumer        = "ReqUnConsumer"
	TypeRespUnConsumer       = "RespUnConsumer"
	TypeReqProduceNormal     = "ReqProduceNormal"
	TypeReqProduceOrdered    = "ReqProduceOrdered"
	TypeReqProduceDelay      = "ReqProduceDelay"
	TypeRespProduceNormal    = "RespProduceNormal"
	TypeRespProduceOrdered   = "RespProduceOrdered"
	TypeRespProduceDelay     = "RespProduceDelay"
	TypeReqConsume           = "ReqConsume"
	TypeRespConsume          = "RespConsume"
	TypeReqConsumeAck        = "ReqConsumeAck"
	TypeRespConsumeAck       = "RespConsumeAck"
	TypeError                = "Error"
)

// 自定义类型替代 []byte
type ByteArray []byte

// 实现 json.Marshaler 接口
func (b ByteArray) MarshalJSON() ([]byte, error) {
	// 将字节切片转换为JSON数组格式
	result := make([]json.Number, len(b))
	for i, v := range b {
		result[i] = json.Number(fmt.Sprintf("%d", v))
	}
	return json.Marshal(result)
}

// Message 顶级消息结构
type Message struct {
	Type string          `json:"type"`
	Data json.RawMessage `json:"data"`
}

// ReqMsgPing 心跳请求
type ReqMsgPing struct {
}

// RespMsgPing 心跳响应
type RespMsgPing struct {
	Msg string `json:"msg"`
}

// ReqMsgSubscriber 订阅请求
type ReqMsgSubscriber struct {
	Topic string `json:"topic"`
}

// RespMsgSubscriber 订阅响应
type RespMsgSubscriber struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
}

// ReqMsgUnsubscriber 取消订阅请求
type ReqMsgUnsubscriber struct {
	Topic string `json:"topic"`
}

// RespMsgUnsubscriber 取消订阅响应
type RespMsgUnsubscriber struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
}

// ReqMsgPublish 发布请求
type ReqMsgPublish struct {
	Topic   string    `json:"topic"`
	Message ByteArray `json:"message"` // 实际使用可能需要Base64编码
}

// RespMsgPublish 发布响应
type RespMsgPublish struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
}

// ReqMsgSubscribe 订阅请求
type ReqMsgSubscribe struct {
	Topic string `json:"topic"`
}

// RespMsgSubscribe 订阅响应
type RespMsgSubscribe struct {
	ID      uint64    `json:"id"`
	Status  MsgStatus `json:"status"`
	Topic   string    `json:"topic"`
	Message ByteArray `json:"message"`
}

// ReqMsgConsumerTopic 消费请求
type ReqMsgConsumerTopic struct {
	Topic string `json:"topic"`
}

// RespMsgConsumerTopic 消费响应
type RespMsgConsumerTopic struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
}

// ReqMsgUnConsumer 取消消费请求
type ReqMsgUnConsumer struct {
	Topic string `json:"topic"`
}

// RespMsgUnConsumer 取消消费响应
type RespMsgUnConsumer struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
}

// ReqMsgProduceNormal 普通生产请求
type ReqMsgProduceNormal struct {
	Topic   string    `json:"topic"`
	Message ByteArray `json:"message"`
}

// RespMsgProduceNormal 普通生产响应
type RespMsgProduceNormal struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
}

// ReqMsgProduceOrdered 有序生产请求
type ReqMsgProduceOrdered struct {
	Topic   string    `json:"topic"`
	Message ByteArray `json:"message"`
}

// RespMsgProduceOrdered 有序生产响应
type RespMsgProduceOrdered struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
}

// ReqMsgProduceDelay 延时生产请求
type ReqMsgProduceDelay struct {
	Topic   string    `json:"topic"`
	Message ByteArray `json:"message"`
	Delay   uint64    `json:"delay"`
}

// RespMsgProduceDelay 延时生产响应
type RespMsgProduceDelay struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Topic  string    `json:"topic"`
	Msg    string    `json:"msg"`
	Delay  uint64    `json:"delay"`
}

type ReqMsgConsume struct {
	Topic string `json:"topic"`
}

type RespMsgConsume struct {
	ID      uint64    `json:"id"`
	Topic   string    `json:"topic"`
	Message ByteArray `json:"message"`
}

// ReqMsgConsumeAck 消费确认请求
type ReqMsgConsumeAck struct {
	ID uint64 `json:"id"`
}

// RespMsgConsumeAck 消费确认响应
type RespMsgConsumeAck struct {
	ID     uint64    `json:"id"`
	Status MsgStatus `json:"status"`
	Msg    string    `json:"msg"`
}

// SerializeMessage 序列化消息为JSON
func Serialize(msg interface{}) ([]byte, error) {
	// 获取消息类型
	var msgType string
	switch msg.(type) {
	case ReqMsgPing:
		msgType = TypeReqPing
	case RespMsgPing:
		msgType = TypeRespPing
	case ReqMsgSubscriber:
		msgType = TypeReqSubscribeTopic
	case RespMsgSubscriber:
		msgType = TypeRespSubscribeTopic
	case ReqMsgUnsubscriber:
		msgType = TypeUnsubscribeTopic
	case RespMsgUnsubscriber:
		msgType = TypeRespUnsubscribeTopic
	case ReqMsgPublish:
		msgType = TypeReqPublish
	case RespMsgPublish:
		msgType = TypeRespPublish
	case ReqMsgSubscribe:
		msgType = TypeReqSubscribe
	case RespMsgSubscribe:
		msgType = TypeRespSubscribe
	case ReqMsgConsumerTopic:
		msgType = TypeReqConsumerTopic
	case RespMsgConsumerTopic:
		msgType = TypeRespConsumerTopic
	case ReqMsgUnConsumer:
		msgType = TypeReqUnConsumer
	case RespMsgUnConsumer:
		msgType = TypeRespUnConsumer
	case ReqMsgProduceNormal:
		msgType = TypeReqProduceNormal
	case RespMsgProduceNormal:
		msgType = TypeRespProduceNormal
	case ReqMsgProduceOrdered:
		msgType = TypeReqProduceOrdered
	case RespMsgProduceOrdered:
		msgType = TypeRespProduceOrdered
	case ReqMsgProduceDelay:
		msgType = TypeReqProduceDelay
	case RespMsgProduceDelay:
		msgType = TypeRespProduceDelay
	case ReqMsgConsume:
		msgType = TypeReqConsume
	case RespMsgConsume:
		msgType = TypeRespConsume
	case ReqMsgConsumeAck:
		msgType = TypeReqConsumeAck
	case RespMsgConsumeAck:
		msgType = TypeRespConsumeAck
	case string: // Error 类型
		msgType = TypeError
	default:
		return nil, fmt.Errorf("unknown message type: %T", msg)
	}

	// 序列化数据部分
	data, err := json.Marshal(msg)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal message data: %w", err)
	}

	// 构建完整消息
	fullMsg := Message{
		Type: msgType,
		Data: data,
	}

	var fullMsgBytes []byte
	fullMsgBytes, err = json.Marshal(fullMsg)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal message envelope: %w", err)
	}

	return fullMsgBytes, nil
}

// MessageFromBytes 从字节数组反序列化消息
func MessageFromBytes(data []byte) (Message, error) {
	var msg Message
	if err := json.Unmarshal(data, &msg); err != nil {
		return Message{}, fmt.Errorf("failed to unmarshal message envelope: %w", err)
	}
	return msg, nil
}

// DeserializeMessage 反序列化JSON消息
func Deserialize(data []byte) (string, interface{}, error) {
	msg, err := MessageFromBytes(data)
	if err != nil {
		return "", nil, fmt.Errorf("failed to parse message: %w", err)
	}

	switch msg.Type {
	case TypeReqPing:
		var req ReqMsgPing
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqPing, req, err
	case TypeRespPing:
		var resp RespMsgPing
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespPing, resp, err
	case TypeReqSubscribeTopic:
		var req ReqMsgSubscriber
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqSubscribeTopic, req, err
	case TypeRespSubscribeTopic:
		var resp RespMsgSubscriber
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespSubscribeTopic, resp, err
	case TypeUnsubscribeTopic:
		var req ReqMsgUnsubscriber
		err := json.Unmarshal(msg.Data, &req)
		return TypeUnsubscribeTopic, req, err
	case TypeRespUnsubscribeTopic:
		var resp RespMsgUnsubscriber
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespUnsubscribeTopic, resp, err
	case TypeReqPublish:
		var req ReqMsgPublish
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqPublish, req, err
	case TypeRespPublish:
		var resp RespMsgPublish
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespPublish, resp, err
	case TypeReqSubscribe:
		var req ReqMsgSubscribe
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqSubscribe, req, err
	case TypeRespSubscribe:
		var resp RespMsgSubscribe
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespSubscribe, resp, err
	case TypeReqConsumerTopic:
		var req ReqMsgConsumerTopic
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqConsumerTopic, req, err
	case TypeRespConsumerTopic:
		var resp RespMsgConsumerTopic
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespConsumerTopic, resp, err
	case TypeReqUnConsumer:
		var req ReqMsgUnConsumer
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqUnConsumer, req, err
	case TypeRespUnConsumer:
		var resp RespMsgUnConsumer
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespUnConsumer, resp, err
	case TypeReqProduceNormal:
		var req ReqMsgProduceNormal
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqProduceNormal, req, err
	case TypeRespProduceNormal:
		var resp RespMsgProduceNormal
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespProduceNormal, resp, err
	case TypeReqProduceOrdered:
		var req ReqMsgProduceOrdered
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqProduceOrdered, req, err
	case TypeRespProduceOrdered:
		var resp RespMsgProduceOrdered
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespProduceOrdered, resp, err
	case TypeReqProduceDelay:
		var req ReqMsgProduceDelay
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqProduceDelay, req, err
	case TypeRespProduceDelay:
		var resp RespMsgProduceDelay
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespProduceDelay, resp, err
	case TypeReqConsume:
		var req ReqMsgConsume
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqConsume, req, err
	case TypeRespConsume:
		var resp RespMsgConsume
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespConsume, resp, err
	case TypeReqConsumeAck:
		var req ReqMsgConsumeAck
		err := json.Unmarshal(msg.Data, &req)
		return TypeReqConsumeAck, req, err
	case TypeRespConsumeAck:
		var resp RespMsgConsumeAck
		err := json.Unmarshal(msg.Data, &resp)
		return TypeRespConsumeAck, resp, err
	case TypeError:
		var errMsg string
		err := json.Unmarshal(msg.Data, &errMsg)
		return TypeError, errMsg, err
	default:
		return "", nil, fmt.Errorf("unknown message type: %s", msg.Type)
	}
}

File: src/lib/message.rs
use std::error::Error;

use bincode::{Decode, Encode};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode, PartialEq, Eq)]
pub enum MessageStatus {
    New,
    Reconsume,
    Pending(u8, u64, u64),
    Dead,
    Acked,
}

impl MessageStatus {
    pub fn is_pending(&self) -> bool {
        match self {
            MessageStatus::Pending(_, _, _) => true,
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct MessageHistory {
    pub status: MessageStatus,
    pub timestamp: u64,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct MessageBox {
    pub id: u64,
    pub status: MessageStatus,
    pub timestamp: u64,
    pub message: Message,
    pub history: Vec<MessageHistory>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Encode, Decode)]
pub enum MsgStatus {
    Success,
    Warning,
    Failure,
    Error,
}

/// Message for Amq.
///
/// # Variants
///
/// - `ReqPing`: Request ping.
/// - `RespPing`: Response ping.
/// - `ReqAuthorizer`: Request authorizer.
/// - `RespAuthorizer`: Response authorizer.
/// - `ReqSubscribeTopic`: Request subscribe topic.
/// - `RespSubscribeTopic`: Response subscribe topic.
/// - `ReqUnsubscribeTopic`: Request unsubscribe topic.
/// - `RespUnsubscribeTopic`: Response unsubscribe topic.
/// - `ReqPublish`: Request publish message.
/// - `RespPublish`: Response publish message.
/// - `ReqSubscribe`: Request subscribe message.
/// - `RespSubscribe`: Response subscribe message.
/// - `ReqConsumerTopic`: Request consumer topic.
/// - `RespConsumerTopic`: Response consumer topic.
/// - `ReqUnconsumerTopic`: Request unconsumer topic.
/// - `RespUnconsumerTopic`: Response unconsumer topic.
/// - `ReqPullMessage`: Request pull message.
/// - `RespPullMessage`: Response pull message.
/// - `ReqProduceNormal`: Request produce normal message.
/// - `ReqProduceOrdered`: Request produce ordered message.
/// - `ReqProduceDelay`: Request produce delay message.
/// - `RespProduceNormal`: Response produce normal message.
/// - `RespProduceOrdered`: Response produce ordered message.
/// - `RespProduceDelay`: Response produce delay message.
/// - `ReqConsume`: Request consume message.
/// - `RespConsume`: Response consume message.
/// - `ReqConsumeAck`: Request consume ack message.
/// - `RespConsumeAck`: Response consume ack message.
/// - `ReqConsumeAckMulti`: Request consume ack multi message.
/// - `RespConsumeAckMulti`: Response consume ack multi message.
/// - `ReqReconsumeLater`: Request reconsume later message.
/// - `RespReconsumeLater`: Response reconsume later message.
/// - `Error`: Error message.
/// - `ReqMessageList`: Request message list.
/// - `RespMessageList`: Response message list.
#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
#[serde(tag = "type", content = "data")]
pub enum Message {
    ReqPing(ReqMsgPing),
    RespPing(RespMsgPing),
    ReqAuthorizer(ReqMsgAuthorizer),
    RespAuthorizer(RespMsgAuthorizer),
    ReqSubscribeTopic(ReqMsgSubscriber),
    RespSubscribeTopic(RespMsgSubscriber),
    ReqUnsubscribeTopic(ReqMsgUnsubscriber),
    RespUnsubscribeTopic(RespMsgUnsubscriber),
    ReqPublish(ReqMsgPublish),
    RespPublish(RespMsgPublish),
    ReqSubscribe(ReqMsgSubscribe),
    RespSubscribe(RespMsgSubscribe),
    ReqConsumerTopic(ReqMsgConsumerTopic),
    RespConsumerTopic(RespMsgConsumerTopic),
    ReqUnconsumerTopic(ReqMsgUnconsumerTopic),
    RespUnconsumerTopic(RespMsgUnconsumerTopic),
    ReqPullMessage(ReqPullMessage),
    RespPullMessage(RespPullMessage),
    ReqProduceNormal(ReqMsgProduceNormal),
    ReqProduceOrdered(ReqMsgProduceOrdered),
    ReqProduceDelay(ReqMsgProduceDelay),
    RespProduceNormal(RespMsgProduceNormal),
    RespProduceOrdered(RespMsgProduceOrdered),
    RespProduceDelay(RespMsgProduceDelay),
    ReqConsume(ReqMsgConsume),
    RespConsume(RespMsgConsume),
    ReqConsumeAck(ReqMsgConsumeAck),
    RespConsumeAck(RespMsgConsumeAck),
    ReqConsumeAckMulti(ReqMsgConsumeAckMulti),
    RespConsumeAckMulti(RespMsgConsumeAckMulti),
    ReqReconsumeLater(ReqReconsumeLater),
    RespReconsumeLater(RespReconsumeLater),
    ReqReconsumeDelay(ReqReconsumeDelay),
    RespReconsumeDelay(RespReconsumeDelay),
    Error(String),
    ReqMessageList(ReqMsgList),
    RespMessageList(RespMsgList),
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgPing {}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgPing {}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgAuthorizer {
    pub access_key: String,
    pub access_secret: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgAuthorizer {
    pub id: u64,
    pub status: MsgStatus,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgSubscriber {
    pub topic: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgSubscriber {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgUnsubscriber {
    pub topic: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgUnsubscriber {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgPublish {
    pub topic: String,
    pub message: Vec<u8>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgPublish {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgSubscribe {
    pub topic: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgSubscribe {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub message: Vec<u8>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgConsumerTopic {
    pub topic: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgConsumerTopic {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgUnconsumerTopic {
    pub topic: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgUnconsumerTopic {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqPullMessage {
    pub topic: String,
    pub total: u32,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespPullMsg {
    pub id: u64,
    pub message: Vec<u8>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespPullMessage {
    pub topic: String,
    pub messages: Vec<RespPullMsg>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgProduceNormal {
    pub topic: String,
    pub message: Vec<u8>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgProduceNormal {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgProduceOrdered {
    pub topic: String,
    pub message: Vec<u8>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgProduceOrdered {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgProduceDelay {
    pub topic: String,
    pub message: Vec<u8>,
    pub delay: u64,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgProduceDelay {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub msg: String,
    pub delay: u64,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgConsume {
    pub topic: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgConsume {
    pub id: u64,
    pub topic: String,
    pub message: Vec<u8>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgConsumeAck {
    pub id: u64,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgConsumeAck {
    pub id: u64,
    pub status: MsgStatus,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgConsumeAckMulti {
    pub ids: Vec<u64>,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgConsumeAckMulti {
    pub id: u64,
    pub status: MsgStatus,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqReconsumeLater {
    pub id: u64,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespReconsumeLater {
    pub id: u64,
    pub status: MsgStatus,
    pub msg: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqReconsumeDelay {
    pub id: u64,
    pub delay: u64,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespReconsumeDelay {
    pub id: u64,
    pub status: MsgStatus,
    pub msg: String,
    pub delay: u64,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct ReqMsgList {
    pub topic: String,
    pub page_size: u32,
    pub page_num: u32,
}

#[derive(Debug, Clone, Deserialize, Serialize, Encode, Decode)]
pub struct RespMsgList {
    pub id: u64,
    pub status: MsgStatus,
    pub topic: String,
    pub message_list: Vec<MessageBox>,
}

impl PartialEq for Message {
    fn eq(&self, _other: &Self) -> bool {
        match (self, _other) {
            (Message::ReqPing(_), Message::ReqPing(_)) => true,
            (Message::RespPing(_), Message::RespPing(_)) => true,
            (Message::ReqAuthorizer(req1), Message::ReqAuthorizer(req2)) => {
                req1.access_key == req2.access_key && req1.access_secret == req2.access_secret
            }
            (Message::RespAuthorizer(resp1), Message::RespAuthorizer(resp2)) => {
                resp1.id == resp2.id && resp1.status == resp2.status && resp1.msg == resp2.msg
            }
            (Message::ReqSubscribeTopic(req1), Message::ReqSubscribeTopic(req2)) => {
                req1.topic == req2.topic
            }
            (Message::RespSubscribeTopic(resp1), Message::RespSubscribeTopic(resp2)) => {
                resp1.id == resp2.id
                    && resp1.status == resp2.status
                    && resp1.topic == resp2.topic
                    && resp1.msg == resp2.msg
            }
            (Message::ReqUnsubscribeTopic(req1), Message::ReqUnsubscribeTopic(req2)) => {
                req1.topic == req2.topic
            }
            (Message::RespUnsubscribeTopic(resp1), Message::RespUnsubscribeTopic(resp2)) => {
                resp1.id == resp2.id
                    && resp1.status == resp2.status
                    && resp1.topic == resp2.topic
                    && resp1.msg == resp2.msg
            }
            (Message::ReqPublish(req1), Message::ReqPublish(req2)) => {
                req1.topic == req2.topic && req1.message == req2.message
            }
            (Message::RespPublish(resp1), Message::RespPublish(resp2)) => {
                resp1.id == resp2.id
                    && resp1.status == resp2.status
                    && resp1.topic == resp2.topic
                    && resp1.msg == resp2.msg
            }
            (Message::ReqSubscribe(req1), Message::ReqSubscribe(req2)) => req1.topic == req2.topic,
            (Message::RespSubscribe(resp1), Message::RespSubscribe(resp2)) => {
                resp1.id == resp2.id
                    && resp1.status == resp2.status
                    && resp1.topic == resp2.topic
                    && resp1.message == resp2.message
            }
            (Message::ReqProduceNormal(req1), Message::ReqProduceNormal(req2)) => {
                req1.topic == req2.topic && req1.message == req2.message
            }
            (Message::RespProduceNormal(resp1), Message::RespProduceNormal(resp2)) => {
                resp1.id == resp2.id
                    && resp1.status == resp2.status
                    && resp1.topic == resp2.topic
                    && resp1.msg == resp2.msg
            }
            (Message::ReqProduceOrdered(req1), Message::ReqProduceOrdered(req2)) => {
                req1.topic == req2.topic && req1.message == req2.message
            }
            (Message::RespProduceOrdered(resp1), Message::RespProduceOrdered(resp2)) => {
                resp1.id == resp2.id
                    && resp1.status == resp2.status
                    && resp1.topic == resp2.topic
                    && resp1.msg == resp2.msg
            }
            (Message::ReqProduceDelay(req1), Message::ReqProduceDelay(req2)) => {
                req1.topic == req2.topic && req1.message == req2.message && req1.delay == req2.delay
            }
            (Message::RespProduceDelay(resp1), Message::RespProduceDelay(resp2)) => {
                resp1.id == resp2.id
                    && resp1.status == resp2.status
                    && resp1.topic == resp2.topic
                    && resp1.msg == resp2.msg
                    && resp1.delay == resp2.delay
            }
            (Message::ReqConsume(req1), Message::ReqConsume(req2)) => req1.topic == req2.topic,
            (Message::RespConsume(resp1), Message::RespConsume(resp2)) => {
                resp1.id == resp2.id && resp1.topic == resp2.topic && resp1.message == resp2.message
            }
            (Message::ReqConsumeAck(req1), Message::ReqConsumeAck(req2)) => req1.id == req2.id,
            (Message::RespConsumeAck(resp1), Message::RespConsumeAck(resp2)) => {
                resp1.id == resp2.id && resp1.status == resp2.status && resp1.msg == resp2.msg
            }
            (Message::ReqReconsumeLater(req1), Message::ReqReconsumeLater(req2)) => {
                req1.id == req2.id
            }
            (Message::RespReconsumeLater(resp1), Message::RespReconsumeLater(resp2)) => {
                resp1.id == resp2.id && resp1.status == resp2.status && resp1.msg == resp2.msg
            }
            (Message::Error(err1), Message::Error(err2)) => err1 == err2,
            _ => false,
        }
    }
}

impl Eq for Message {}

impl Message {
    pub fn serialize(&self) -> Result<Vec<u8>, Box<dyn Error>> {
        let bytes = serde_json::to_vec(self)?;
        Ok(bytes)
    }

    pub fn deserialize(bytes: &[u8]) -> Result<Self, Box<dyn Error>> {
        let message: Message = serde_json::from_slice(bytes)?;
        Ok(message)
    }
}
